File: Python/PoetryTest/src/poetrytest/sub_module/redis_client_example.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import logging
import redis
import uuid
from typing import Any, Awaitable
from redis.typing import ResponseT


# pyright会报ResponseT的类型错误，使用此函数包装
def extract(value: ResponseT) -> Any:
    if isinstance(value, Awaitable):
        raise TypeError
    return value


# pyright会报ResponseT的类型错误，使用此函数包装
async def extract_async(value: ResponseT) -> Any:
    if not isinstance(value, Awaitable):
        raise TypeError
    return await value


def main():
    logging.basicConfig(
        level=logging.DEBUG,
        format="[%(asctime)s] %(levelname)s [%(thread)s][%(filename)s:%(lineno)d] - %(message)s",
    )

    # 建立连接
    # socket_connect_timeout有issues，不生效
    try:
        connect_timeout: float = 2.0
        redis_cache = redis.Redis(
            host='localhost',
            port=6379,
            # username='dvora',
            # password='redis',
            decode_responses=True,
            socket_timeout=connect_timeout,
            socket_connect_timeout=connect_timeout,
            retry_on_timeout=False,
            socket_keepalive=True,
        )
        redis_cache.ping()
    except Exception as e:
        logging.error(f"Redis连接失败: {e}", exc_info=True)
        return

    # 设定Key
    key: str = "PythonKey"
    uuid_str: str = str(uuid.uuid4())
    set_value: str = "使用Python客户端添加_" + uuid_str
    redis_cache.set(key, set_value)
    logging.info(f"添加Key成功    key={key}    value={set_value}")

    # 取得Key
    get_value: str = extract(redis_cache.get(key))
    logging.info(f"取得Key成功    key={key}    value={get_value}")

    # 判断Key是否存在
    is_exists: int = extract(redis_cache.exists(key))
    logging.info(f"Key[{key}]是否存在    {is_exists}")
    not_existkey: str = "abcde"
    is_exists: int = extract(redis_cache.exists(not_existkey))
    logging.info(f"Key[{not_existkey}]是否存在    {is_exists}")

    # 为Key设置过期时间(秒)
    expire_seconds: int = 10
    redis_cache.expire(key, expire_seconds)
    logging.info(f"设定Key[{key}]的过期时间为：{expire_seconds}秒")

    # 关闭连接
    redis_cache.close()


if __name__ == "__main__":
    main()

File: Web/TSSampleProject/src/RedisClientExample.ts
import { createClient, RedisClientType } from 'redis';
import { v4 as uuidv4 } from 'uuid';
import { getLogger } from './log/log.js';

// 日志
const log = getLogger();

// 异步开始函数
async function startRedisAsync(): Promise<void> {

	// 连接参数
	//const client: RedisClientType = await createClient();
	const client: RedisClientType = await createClient({
		socket: {
			host: 'localhost',
			port: 6379,
			connectTimeout: 1000,
			reconnectStrategy: retries => false,
		}
	});

	// 建立连接
	client.on('error', err => {
		log.error('Redis连接失败');
		log.error(err);
		throw err;
	}).connect();

	// 设定Key
	const key: string = 'TSKey';
	const uuid: string = uuidv4();
	const setValue: string = '使用Node客户端添加_' + uuid;
	await client.set(key, setValue);
	//console.log(`添加Key成功    key=${key}    value=${setValue}`);
	log.info(`添加Key成功    key=${key}    value=${setValue}`);

	// 取得Key
	const getValue: string = await client.get(key) || '';
	log.info(`取得Key成功    key=${key}    value=${getValue}`);

	// 判断Key是否存在
	let isExists: number = await client.exists(key);
	log.info(`Key[${key}]是否存在    ${isExists}`);
	const notExistkey: string = 'abcde';
	isExists = await client.exists(notExistkey);
	log.info(`Key[${notExistkey}]是否存在    ${isExists}`);

	// 为Key设置过期时间(秒)
	const expireSeconds: number = 10;
	await client.expire(key, expireSeconds);
	log.info(`设定Key[${key}]的过期时间为：${expireSeconds}秒`);

	//await client.quit();
	await client.disconnect();
	log.info("已断开Redis连接");
}

// 主函数
(async function main() {
	try {
		await startRedisAsync();
	} catch (e) {
		log.error(e);
	} finally {
		// 因为是异步日志，所以要在最后刷新
		log.flush();
		// 结束
		process.exit(0);
	}
})();
